File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/flutter/AbrBridge.kt
package com.theoplayer.flutter

import com.theoplayer.android.api.player.Player
import com.theoplayer.android.api.abr.AbrStrategyConfiguration
import com.theoplayer.android.api.abr.AbrStrategyMetadata
import com.theoplayer.android.api.abr.AbrStrategyType
import com.theoplayer.flutter.pigeon.AbrStrategyConfigurationPigeon
import com.theoplayer.flutter.pigeon.AbrStrategyMetadataPigeon
import com.theoplayer.flutter.pigeon.AbrStrategyTypePigeon
import com.theoplayer.flutter.pigeon.THEOplayerNativeAbrAPI

class AbrBridge(
    private val pigeonMessenger: PigeonBinaryMessengerWrapper,
    private val player: Player
) : THEOplayerNativeAbrAPI {

    init {
        THEOplayerNativeAbrAPI.setUp(pigeonMessenger, this)
    }

    fun dispose() {
        THEOplayerNativeAbrAPI.setUp(pigeonMessenger, null)
    }

    // MARK: - THEOplayerNativeAbrAPI

    override fun getAbrStrategy(): AbrStrategyConfigurationPigeon {
        val native = player.abr.abrStrategy
        return AbrStrategyConfigurationPigeon(
            type = toPigeonType(native.type),
            metadata = native.metadata?.let { meta ->
                AbrStrategyMetadataPigeon(bitrate = meta.bitrate?.toLong())
            }
        )
    }

    override fun setAbrStrategy(config: AbrStrategyConfigurationPigeon) {
        val metadata = config.metadata?.bitrate?.let { bitrate ->
            AbrStrategyMetadata(bitrate = bitrate.toInt())
        }
        player.abr.abrStrategy = AbrStrategyConfiguration(
            type = toNativeType(config.type),
            metadata = metadata
        )
    }

    override fun getTargetBuffer(): Double {
        return player.abr.targetBuffer?.toDouble() ?: 20.0
    }

    override fun setTargetBuffer(value: Double) {
        player.abr.targetBuffer = value.toInt()
    }

    override fun getPreferredPeakBitRate(): Double {
        // Not supported on Android
        return 0.0
    }

    override fun setPreferredPeakBitRate(value: Double) {
        // No-op on Android
    }

    // MARK: - Mapping helpers

    private fun toPigeonType(type: AbrStrategyType): AbrStrategyTypePigeon {
        return when (type) {
            AbrStrategyType.PERFORMANCE -> AbrStrategyTypePigeon.PERFORMANCE
            AbrStrategyType.QUALITY -> AbrStrategyTypePigeon.QUALITY
            AbrStrategyType.BANDWIDTH -> AbrStrategyTypePigeon.BANDWIDTH
        }
    }

    private fun toNativeType(type: AbrStrategyTypePigeon): AbrStrategyType {
        return when (type) {
            AbrStrategyTypePigeon.PERFORMANCE -> AbrStrategyType.PERFORMANCE
            AbrStrategyTypePigeon.QUALITY -> AbrStrategyType.QUALITY
            AbrStrategyTypePigeon.BANDWIDTH -> AbrStrategyType.BANDWIDTH
        }
    }
}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/AbrBridge.swift
//
//  AbrBridge.swift
//
//  Bridge between Flutter Pigeon API and iOS ABRConfiguration.
//

import Foundation
import Flutter
import THEOplayerSDK

class AbrBridge: THEOplayerNativeAbrAPI {

    private let pigeonMessenger: PigeonBinaryMessengerWrapper
    private let theoplayer: THEOplayer

    init(theoplayer: THEOplayer, pigeonMessenger: PigeonBinaryMessengerWrapper) {
        self.theoplayer = theoplayer
        self.pigeonMessenger = pigeonMessenger
        THEOplayerNativeAbrAPISetup.setUp(binaryMessenger: pigeonMessenger, api: self)
    }

    func dispose() {
        THEOplayerNativeAbrAPISetup.setUp(binaryMessenger: pigeonMessenger, api: nil)
    }

    // MARK: - THEOplayerNativeAbrAPI

    func getAbrStrategy() throws -> AbrStrategyConfigurationPigeon {
        let native = theoplayer.abr.strategy
        return AbrStrategyConfigurationPigeon(
            type: toPigeonType(native.type),
            metadata: native.metadata.map { meta in
                AbrStrategyMetadataPigeon(bitrate: meta.bitrate.map { Int64($0) })
            }
        )
    }

    func setAbrStrategy(config: AbrStrategyConfigurationPigeon) throws {
        let nativeType = toNativeType(config.type)
        var nativeMetadata: ABRMetadata? = nil
        if let bitrate = config.metadata?.bitrate {
            nativeMetadata = ABRMetadata(bitrate: Double(bitrate))
        }
        theoplayer.abr.strategy = ABRStrategyConfiguration(type: nativeType, metadata: nativeMetadata)
    }

    func getTargetBuffer() throws -> Double {
        return theoplayer.abr.targetBuffer ?? 20.0
    }

    func setTargetBuffer(value: Double) throws {
        theoplayer.abr.targetBuffer = value
    }

    func getPreferredPeakBitRate() throws -> Double {
        return theoplayer.abr.preferredPeakBitRate ?? 0.0
    }

    func setPreferredPeakBitRate(value: Double) throws {
        theoplayer.abr.preferredPeakBitRate = value
    }

    // MARK: - Mapping helpers

    private func toPigeonType(_ type: ABRStrategyType) -> AbrStrategyTypePigeon {
        switch type {
        case .performance: return .performance
        case .quality: return .quality
        case .bandwidth: return .bandwidth
        }
    }

    private func toNativeType(_ type: AbrStrategyTypePigeon) -> ABRStrategyType {
        switch type {
        case .performance: return .performance
        case .quality: return .quality
        case .bandwidth: return .bandwidth
        }
    }
}
